File: 프로그래머스/lv3/12971. 스티커 모으기（2）/스티커 모으기（2）.py
def solution(sticker):
    answer = 0
    if len(sticker) == 1:
        return sticker[0]
    if len(sticker) == 2:
        return max(sticker[0], sticker[1])
    dp = [0 for _ in range(len(sticker)-1)]
    dp_ = [0 for _ in range(len(sticker))]
    dp[0] = sticker[0]    
    dp[1] = max(dp[0], sticker[1])
    dp_[1] = sticker[1]
    dp_[2] = max(dp_[1], sticker[2])
    
    for i in range(2, len(sticker)-1):
        dp[i] = max(dp[i-2] + sticker[i], dp[i-1])
    for i in range(3, len(sticker)):
        dp_[i] = max(dp_[i-2] + sticker[i], dp_[i-1])
                
    return max(dp[-1], dp_[-1])

File: 프로그래머스/3/12971. 스티커 모으기（2）/스티커 모으기（2）.swift
func solution(_ sticker: [Int]) -> Int {
    let n = sticker.count
    
    if n == 1 {
        return sticker[0]
    }
    if n == 2 {
        return max(sticker[0], sticker[1])
    }
    
    var dp1 = [Int](repeating: 0, count: n - 1)
    dp1[0] = sticker[0]
    dp1[1] = max(sticker[0], sticker[1])
    
    var dp2 = [Int](repeating: 0, count: n)
    dp2[0] = 0
    dp2[1] = sticker[1]
    dp2[2] = max(sticker[1], sticker[2])
    
    for i in 2..<n - 1 {
        dp1[i] = max(dp1[i - 2] + sticker[i], dp1[i - 1])
        dp2[i + 1] = max(dp2[i - 1] + sticker[i + 1], dp2[i])
    }
    
    return max(dp1[n - 2], dp2[n - 1])
}
